### PR TITLE
raftstore: allow follower read with applied index

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

###  What have you changed?

Allow follower read with applied index. So that TiKV can response the request if it applys to the given position, without any communication with the region's leader.

###  What is the type of the changes?

Improvement.

###  How is the PR tested?
Please select the tests that you ran to verify your changes:
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No.

###  Does this PR affect `tidb-ansible`?
No.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -116,7 +116,7 @@ prost = "0.5.0"
 regex = "1.3"
 derive-new = "0.5"
 tipb = { git = "https://github.com/pingcap/tipb.git" }
-kvproto = { git = "https://github.com/pingcap/kvproto.git" }
+kvproto = { git = "https://github.com/hicqu/kvproto.git", branch = "follower-read-with-applied" }
```
[the remark this answers is not in the document]
Comment (hicqu) [reply Cargo.toml]: Thanks. I will update it after the kvproto PR is merged.

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -134,6 +137,7 @@ impl ReadIndexQueue {
                 "cannot find corresponding read from pending reads";
                 "id"=>?id, "read-index" =>read_index,
             );
+            panic!("advance read index fail");
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/raftstore/store/peer.rs]: Nice catch!

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -283,6 +311,9 @@ pub struct Peer {
     leader_missing_time: Option<Instant>,
     leader_lease: Lease,
     pending_reads: ReadIndexQueue,
+    // Some requests carry an applied index, so it's not necessary to call `read_index`.
+    // Just wait the Raft applys to the given index and then response those requests.
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: applying

**Comment by siddontang** [comment]
how to use this feature in TiDB?

**Comment by hicqu** [comment]
@siddontang The TiDB side is in developing. The key is TiDB fetches committed indices for target regions, and then send read requests with those indices to TiKVs.

**Pull request closed without merging**